File: packages/python/mythos_sdk/api_client.py
import uuid
from dataclasses import dataclass
from typing import Any
from urllib.parse import quote

import httpx

from .config import load_config
from .errors import (
    InsufficientFundsError,
    InvalidUsageError,
    MythosUnreachableError,
    MythosUpstreamError,
    SessionExpiredError,
    SessionNotFoundError,
)
from .http import get_http_client
@dataclass(frozen=True)
class MeterResult:
    charge_id: str
    session_metered_total: int | None


def read_identity_fields(body: object) -> tuple[str, str | None] | None:
    if not isinstance(body, dict):
        return None
    data = body.get("data")
    if not isinstance(data, dict):
        return None
    token = data.get("llm_identity_token")
    if not isinstance(token, str) or not token:
        return None
    expires_at = data.get("llm_identity_expires_at")
    return token, expires_at if isinstance(expires_at, str) else None


def _encode_jti(jti: str) -> str:
    return quote(jti, safe="")


def _validate_credits(credits: int) -> None:
    if not isinstance(credits, int) or isinstance(credits, bool) or credits <= 0:
        raise InvalidUsageError("credits must be a positive integer")


async def consume_session(jti: str) -> Any:
    config = load_config()
    client = get_http_client()
    return await client.post(
        f"{config.api_url}/api/apps/sessions/{_encode_jti(jti)}/consume",
        json={},
    )


async def meter_session(
    jti: str,
    credits: int,
    reason: str | None = None,
    charge_id: str | None = None,
) -> MeterResult:
    _validate_credits(credits)
    config = load_config()
    resolved_charge_id = charge_id if charge_id is not None else str(uuid.uuid4())
    body: dict[str, Any] = {"credits": credits, "charge_id": resolved_charge_id}
    if reason is not None:
        body["reason"] = reason

    client = get_http_client()
    try:
        resp = await client.post(
            f"{config.api_url}/api/apps/sessions/{_encode_jti(jti)}/meter",
            json=body,
        )
    except httpx.HTTPError as err:
        raise MythosUnreachableError(f"Could not reach Mythos API: {err}") from err

    if resp.status_code == 402:
        raise InsufficientFundsError()
    code: str | None = None
    if not (200 <= resp.status_code < 300):
        try:
            response_body = resp.json()
            if isinstance(response_body, dict) and isinstance(response_body.get("code"), str):
                code = response_body["code"]
        except (ValueError, TypeError):
            pass
        if code in {"SESSION_EXPIRED", "SESSION_NOT_STARTED"}:
            raise SessionExpiredError()
        if resp.status_code == 404:
            raise SessionNotFoundError(jti)
        raise MythosUpstreamError("Meter request failed", resp.status_code, code)

    try:
        response_body = resp.json()
    except (ValueError, TypeError):
        response_body = None
    data = response_body.get("data") if isinstance(response_body, dict) else None
    total = data.get("session_metered_total") if isinstance(data, dict) else None
    session_total = total if isinstance(total, int) and not isinstance(total, bool) and total >= 0 else None
    return MeterResult(charge_id=resolved_charge_id, session_metered_total=session_total)

File: packages/node/src/api-client.ts
import { randomUUID } from 'crypto';
import { mythosRequest } from './http';
import {
  InsufficientFundsError,
  InvalidUsageError,
  MythosUnreachableError,
  MythosUpstreamError,
  SessionExpiredError,
  SessionNotFoundError,
} from './errors';
import type { MythosSession } from './types';

function encodeJti(jti: string): string {
  return encodeURIComponent(jti);
}

function validateCredits(credits: number): void {
  if (!Number.isInteger(credits) || credits <= 0) {
    throw new InvalidUsageError('credits must be a positive integer');
  }
}

async function post(path: string, body: unknown): Promise<Response> {
  return mythosRequest(path, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

export async function readErrorCode(res: Response): Promise<string | undefined> {
  try {
    const body: unknown = await res.json();
    if (isRecord(body) && typeof body['code'] === 'string') return body['code'];
  } catch {
    // A non-JSON error response has no machine-readable code.
  }
  return undefined;
}

export interface MeterResult {
  chargeId: string;
  sessionMeteredTotal: number | null;
}

export function readIdentityFields(
  body: unknown,
): Pick<MythosSession, 'llmIdentityToken' | 'llmIdentityExpiresAt'> | null {
  if (!isRecord(body) || !isRecord(body['data'])) return null;
  const token = body['data']['llm_identity_token'];
  if (typeof token !== 'string' || token.length === 0) return null;
  const expiresAt = body['data']['llm_identity_expires_at'];
  return {
    llmIdentityToken: token,
    ...(typeof expiresAt === 'string' ? { llmIdentityExpiresAt: expiresAt } : {}),
  };
}

export async function consumeSession(jti: string): Promise<Response> {
  return post(`/api/apps/sessions/${encodeJti(jti)}/consume`, {});
}

export async function meterSession(
  jti: string,
  credits: number,
  reason?: string,
  chargeId?: string,
): Promise<MeterResult> {
  validateCredits(credits);
  const resolvedChargeId = chargeId ?? randomUUID();

  const body: Record<string, unknown> = {
    credits,
    charge_id: resolvedChargeId,
  };
  if (reason !== undefined) {
    body.reason = reason;
  }

  let res: Response;
  try {
    res = await post(`/api/apps/sessions/${encodeJti(jti)}/meter`, body);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new MythosUnreachableError(`Could not reach Mythos API: ${message}`);
  }

  if (res.status === 402) throw new InsufficientFundsError();
  if (!res.ok) {
    const code = await readErrorCode(res);
    if (code === 'SESSION_EXPIRED' || code === 'SESSION_NOT_STARTED') throw new SessionExpiredError();
    if (res.status === 404) throw new SessionNotFoundError(jti);
    throw new MythosUpstreamError('Meter request failed', res.status, code);
  }

  let responseBody: unknown;
  try {
    responseBody = await res.json();
  } catch {
    responseBody = null;
  }
  const data = isRecord(responseBody) && isRecord(responseBody['data']) ? responseBody['data'] : null;
  const total = data?.['session_metered_total'];
  return {
    chargeId: resolvedChargeId,
    sessionMeteredTotal: typeof total === 'number' && Number.isSafeInteger(total) ? total : null,
  };
}
